analyzer/analyzer.go

```diff
@@ -30,310 +23,131 @@ var (
+}
 
-	layerInfo := types.BlobInfo{
-		Digest:        layerDigest,
-		DiffID:        diffID,
-		SchemaVersion: types.BlobJSONSchemaVersion,
-		OS:            os,
-		PackageInfos:  pkgs,
-		Applications:  apps,
-		OpaqueDirs:    opqDirs,
-		WhiteoutFiles: whFiles,
-	}
-	return layerInfo, nil
+func (r *AnalysisResult) isEmpty() bool {
+	return r.OS == nil && len(r.PackageInfos) == 0 && len(r.Applications) == 0
 }
 
-func (ac Config) analyzeConfig(imageID string, osFound types.OS) error {
-	configBlob, err := ac.Extractor.ConfigBlob()
-	if err != nil {
-		return xerrors.Errorf("unable to get config blob: %w", err)
+func (r *AnalysisResult) Merge(new AnalysisResult) {
+	if new.isEmpty() {
+		return
 	}
 
-	// special file for config
-	files := extractor.FileMap{
-		"/config": configBlob,
-	}
-	pkgs, _ := GetPackagesFromCommands(osFound, files)
+	// this struct is accessed by multiple goroutines
+	r.m.Lock()
+	defer r.m.Unlock()
 
-	var s1 v1.ConfigFile
-	if err := json.Unmarshal(configBlob, &s1); err != nil {
-		return xerrors.Errorf("json marshal error: %w", err)
+	if new.OS != nil {
+		r.OS = new.OS
 	}
 
-	info := types.ArtifactInfo{
-		SchemaVersion:   types.ArtifactJSONSchemaVersion,
-		Architecture:    s1.Architecture,
-		Created:         s1.Created.Time,
-		DockerVersion:   s1.DockerVersion,
-		OS:              s1.OS,
-		HistoryPackages: pkgs,
+	if len(new.PackageInfos) > 0 {
+		r.PackageInfos = append(r.PackageInfos, new.PackageInfos...)
 	}
 
-	if err := ac.Cache.PutArtifact(imageID, info); err != nil {
-		return xerrors.Errorf("failed to put image info into the cache: %w", err)
+	if len(new.Applications) > 0 {
+		r.Applications = append(r.Applications, new.Applications...)
 	}
-
-	return nil
 }
 
-type Applier struct {
-	cache cache.LocalArtifactCache
-}
-
-func NewApplier(c cache.LocalArtifactCache) Applier {
-	return Applier{cache: c}
-}
-
-func (a Applier) ApplyLayers(imageID string, diffIDs []string) (types.ArtifactDetail, error) {
-	var layers []types.BlobInfo
-	for _, diffID := range diffIDs {
-		layer, _ := a.cache.GetBlob(diffID)
-		if layer.SchemaVersion == 0 {
-			return types.ArtifactDetail{}, xerrors.Errorf("layer cache missing: %s", diffID)
+func AnalyzeFile(filePath string, info os.FileInfo, opener Opener) (AnalysisResult, error) {
+	var result AnalysisResult
+	for _, analyzer := range analyzers {
+		// filepath extracted from tar file doesn't have the prefix "/"
+		if !analyzer.Required(strings.TrimLeft(filePath, "/"), info) {
+			continue
+		}
+		b, err := opener()
+		if err != nil {
+			return AnalysisResult{}, xerrors.Errorf("unable to open a file (%s): %w", filePath, err)
 		}
-		layers = append(layers, layer)
-	}
-
-	mergedLayer := docker.ApplyLayers(layers)
-	if mergedLayer.OS == nil {
-		return mergedLayer, ErrUnknownOS // send back package and apps info regardless
-	} else if mergedLayer.Packages == nil {
-		return mergedLayer, ErrNoPkgsDetected // send back package and apps info regardless
-	}
```

Comment:
What happens in this use case now?

Reply:
Oh I guess this is reintroduced here? https://github.com/aquasecurity/fanal/pull/111/files?file-filters%5B%5D=.go#diff-8d7f91425a215d1c8e3d51d7b49ebfa2R28-R33 

Am I correct?

Reply:
Right. `ApplyLayers` was in `analyzer.go`, but I moved it in the dedicated file, `applier.go`.